Refactor PluginMenuContributionHandler

colin-grant-work opened the pull request:
- Adds support for `when` contexts in submenus
 - Adds support for treating menu contributions as toolbar contributions

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

Fixes #11264 by introducing support for `when` clauses in submenu contributions.

Also refactors the `PluginMenuContributionHandler`.

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

> In progress: I intend to write a plugin that tests the various menu contribution points we support.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

msujew reviewed (commented):
Looks quite good, at least all menu contribution points mentioned in the testing steps seem to work as expected with the provided test extension.

By the way, while installing Gitlens as a test for this, I noticed that webviews don't display for me in web environments for some reason. I always get a 404 not found error while loading them, even fairly old versions of Theia (tested with 1.26.0, 1.25.0 and 1.20.0). Can you reproduce this issue @colin-grant-work? I didn't want to open a separate issue for this, if it only related to my environment.

colin-grant-work commented:
@msujew, thanks for taking a look. I'm currently working on a more thorough refactor to support submenu registration better without forcing all plugin items into a single group, since that isn't the behavior intended by VSCode. I'll address your comments - and remove all of my random console logs :-) - as I work through that.

colin-grant-work commented on packages/plugin-ext/src/main/browser/menus/plugin-menu-command-adapter.ts:
```diff
@@ -0,0 +1,256 @@
+        const arg = args[0];
+        timelineArgs.push(this.toTimelineArg(arg));
+        timelineArgs.push(CodeUri.parse(arg.uri));
+        timelineArgs.push('source' in arg ? arg.source : '');
```
One good thing about moving code around is that it gives reviewers a chance to revisit the old code and propose improvements. :-).

colin-grant-work commented on packages/plugin-ext/src/main/browser/menus/vscode-theia-menu-mappings.ts:
```diff
@@ -0,0 +1,85 @@
+]);
+
+type CodeEditorWidget = EditorWidget | WebviewWidget;
+@injectable()
+export class CodeEditorWidgetUtil {
```
I'd agree, but this class already existed in this form - I just moved it out of the `plugin-menus-contribution-handler` file. Happy to deprecate / remove in favor of a namespace, though.

colin-grant-work commented on packages/plugin-ext/src/main/browser/menus/plugin-menu-command-adapter.ts:
```diff
@@ -0,0 +1,256 @@
+
+    /* eslint-disable @typescript-eslint/no-explicit-any */
+
+    protected toCommentArgs(...args: any[]): any[] {
```
All of these argument converters are old code moved out of the `MenusContributionPointHandler` class, and the code usually has fairly strong assumptions about what the data will look like (not `undefined`, some kind of `object`), so converting the declarations to `unknown` would require a bunch of extra type checks on types where I don't actually really know what's coming in - hopefully the original implementer did. I can go through and do the checks and ensure things have the shape we expect  - I agree that this code is fairly brittle if anything changes - but since it is all copied over, I think I'd prefer to put that in a separate PR.

colin-grant-work commented:
@planger, would you mind taking a look at the Playwright failures here? The failures have to do with menus, and the PR has to do with menus, so it seems likely that the changes have caused the failures, but I took a bit of a look and couldn't figure out where things are going wrong: the menus that the tests claim to be looking for are present, and the PR doesn't change any of the HTML rendering that they hook onto, as far as I can tell.

planger commented:
@colin-grant-work Sure, I'm happy to have a look! I'm currently on vacation though (including next week), but can look into it right after. Is that soon enough? If it can't wait I'll ask someone in our team to have a look.

colin-grant-work commented:
@planger, I was able to track down the problem. My code had added trailing separators to menus, and Phosphor was automatically hiding those using a `p-mod-collapsed` class. That was preventing the `label()` method from returning, as the selector was never visible. I have added handling to the `MenuItem` class that checks for that class name and returns empty for its label and its shortcut if that class is found. That way, the code can handle invisible items but correctly treats it as empty / invisible for any checks of the item's status. But I didn't mean to have trailing separators, so I've also removed those in a separate commit.

vince-fugnitto reviewed (approved):
LGTM 👍 I confirmed the latest changes work well.
